cumulus_library/statistics/psm.py

```diff
@@ -0,0 +1,247 @@
+    """
+
+    classification_json: str
+    pos_source_table: str
+    neg_source_table: str
+    target_table: str
+    primary_ref: str
+    count_ref: str
+    count_table: str
+    dependent_variable: str
+    pos_sample_size: int
+    neg_sample_size: int
+    join_cols_by_table: dict[str, dict]
+    seed: int
+
+
+class PsmBuilder(BaseTableBuilder):
+    """TableBuilder for creating PSM tables"""
+
+    display_text = "Building PSM tables..."
+
+    def __init__(self, toml_config_path: str):
+        """Loads PSM job details from a PSM configuration file"""
+        with open(toml_config_path, encoding="UTF-8") as file:
+            toml_config = toml.load(file)
+        self.config = PsmConfig(
+            classification_json=f"{PosixPath(toml_config_path).parent}/{toml_config['classification_json']}",
+            pos_source_table=toml_config["pos_source_table"],
+            neg_source_table=toml_config["neg_source_table"],
+            target_table=toml_config["target_table"],
+            primary_ref=toml_config["primary_ref"],
+            dependent_variable=toml_config["dependent_variable"],
+            pos_sample_size=toml_config["pos_sample_size"],
+            neg_sample_size=toml_config["neg_sample_size"],
+            join_cols_by_table=toml_config.get("join_cols_by_table", {}),
+            count_ref=toml_config.get("count_ref", None),
+            count_table=toml_config.get("count_table", None),
+            seed=toml_config.get("seed", 123),
+        )
+        super().__init__()
+
+    def _get_symptoms_dict(self, path: str) -> dict:
+        """convenience function for loading symptoms dictionaries from a json file"""
+        with open(path) as f:
+            symptoms = json.load(f)
+        return symptoms
+
+    def _get_sampled_ids(
+        self,
+        cursor: DatabaseCursor,
+        schema: str,
+        query: str,
+        sample_size: int,
+        dependent_variable: str,
+        is_positive: bool,
+    ):
+        """Creates a table containing randomly sampled patients for PSM analysis
+
+        To use this, it is assumed you have already identified a cohort of positively
+        IDed patients as a manual process.
+        :param cursor: A valid DatabaseCusror:
+        :param schema: the schema/database name where the data exists
+        :param query: a query generated from the psm_dsitinct_ids template
+        :param sample_size: the number of records to include in the random sample.
+            This should generally be >= 20.
+        :param dependent_variable: the name to use for your filtering column
+        :param is_positive: defines the value to be used for your filtering column
+        """
+        df = cursor.execute(query).as_pandas()
+        df = (
+            df.sort_values(by=[self.config.primary_ref])
+            .reset_index()
+            .drop("index", axis=1)
+        )
+        df = (
+            # TODO: remove replace behavior after increasing data sample size
```

Comment:
Is this still valid? As a comment to a non-Matt, I'm not sure it's clear to me when that would be true / why replace is bad. I would currently guess this is about unit tests. (Same comment above the call to `knn_matched` later)